specs/transport/car/carv1/index.md

```diff
@@ -89,7 +90,17 @@ Immediately following the Header block, **one or more** IPLD blocks are concaten
 
 #### Length
 
-Each Section begins with a varint representation of an unsigned integer indicating the number of bytes containing the remainder of the section.
+Each Section begins with a unsigned varint representation of an unsigned integer indicating the number of bytes containing the remainder of the section. The length must be an integer greater than zero. Encountering a zero in the length position of a new _Section_ should be considered an error when reading a CAR. Likewise, a CAR writer should only write non-zero length fields.
+
+A CAR is proplerly terminated when an end-of-stream is encountered when attempting to read a new _Section_.
+
+##### Optional mode: null byte as EOF
+
+By default a CAR reader and writer should not support the null byte (`0x00`, varint value of `0`) at the beginning of a _Section_; each _Section_ should begin with a valid unsigned varint of value greater than zero. However, a system using the CAR format may optionally choose to allow the null byte to signal termination of a CAR, allowing it to optionally be followed by arbitrary bytes that are will not considered part of the CAR.
```

Comment:
@masih think of filecoin as a pallet-freight service, that offers different pallet sizes each being exactly 2^N large, (with 1 byte out of every 128 lost for "margins")

What you put on the pallet is up to you ( .car file in this case ), the rest of the space is "air" or `\x00`. On read the go-car reader is given the entire pallet, `\x00`s included: in a "deal with it" fashion.

Reply:
@willscott made an argument on the IPLD call to consider the null padding as part of the CARv1 file itself.

Personally I don't have strong feelings here, but one reason to agree with him is to have a common understanding that the null padding should be allowed anywhere that accepts a CARv1. That would go against the wording here that decoders shouldn't allow the padding by default, though.

Something to note is what we should call the chunks of bytes that Filecoin are doing proofs on. As per the writing of this spec, it can't be simply a CARv1, because the padding is not part of a CARv1 file. But on the other hand, if we say that a deterministic CARv1 should have zero padding, then they're not using a deterministic CARv1 either.

I think, whatever we do, we have to say that a deterministic CARv1 has no padding, and that Filecoin does proofs over a deterministic CARv1 *plus* padding.

Then whether or not a padding in a non-deterministic CARv1 is part of the file is up to terminology and docs.

Reply:
That's really helpful thanks @ribasushi; would you say then that the "air" is indeed part of the CAR, or it is not and the "air" is specific to FileCoin?

If FileCoin does proof on CARv1 + padding, then I think at the spec level it makes more sense to say padding is indeed part of the CARv1, since we have decided that fixing this in FileCoin itself is out of scope.

Reply:
@masih filecoin (proofs/chain-mechanics/etc) does not care what you put in the deal (container). It is go-fil-markets that does. Since go-fil-markets uses go-car to parse out what's in the container that filecoin provided - you have to accommodate based on that.

The "air" is only "air" because go-fil-markets doesn't usually have enough data to fil a container "perfectly".

Reply:
> Something to note is what we should call the chunks of bytes that Filecoin are doing proofs on

@mvdan It's called a `piece`, and is in the filecoin spec.
A `piece` is always exactly 2**N bytes large, and provides exactly (2**(N-1))*127 bytes of "payload capacity".
See previous answer to how the `piece` is interpreted

Reply:
Thanks for pointing this out; the terminology I am using in this case then is not accurate.

I do think it would be simpler to comprehend the rationale behind null padding's existence and easier to use CARs, may it be in go-fil-market or elsewhere, if null padding is a feature of CARv1 itself in the spec.

Reply:
By the time Filecoin is at the proofs stage it's just an opaque blob of data. It's not even a CAR because it's not only got this null padding at the end but it's got 2 zero bits added for every 256 bits. The null byte thing is only being used because we're undoing both of these padding operations but we apparently haven't retained the length of the source so can't just slice off those zeros from the end.

Reply:
I see in that case either works for me. It's such a shame this needs to be added here in the specs of CARv1 rather than fixed at root in filecoin but hey.